extensions/redis-cache/src/main/java/io/druid/client/cache/RedisCache.java

```diff
@@ -0,0 +1,315 @@
+{
+  private static final Logger log = new Logger(RedisCache.class);
+
+  private final Supplier<ResourceHolder<ShardedJedis>> supplier;
+  private final String prefix;
+  private final long expiration;
+  private final RedisTranscoder transcoder;
+  private final AtomicLong hitCount = new AtomicLong(0);
+  private final AtomicLong missCount = new AtomicLong(0);
+  private final AtomicLong timeoutCount = new AtomicLong(0);
+  private final AtomicLong errorCount = new AtomicLong(0);
+
+  public static Cache create(final RedisCacheConfig config)
+  {
+    final Supplier<ResourceHolder<ShardedJedis>> clientSupplier;
+
+    if (config.getPoolSize() > 1) {
+      clientSupplier = new LoadBalancingPool<>(
+          config.getPoolSize(),
+          new Supplier<ShardedJedis>()
+          {
+            @Override
+            public ShardedJedis get()
+            {
+              return new ShardedJedis(config.getShardsInfo(), Hashing.MURMUR_HASH);
+            }
+          }
+      );
+    } else {
+      clientSupplier = Suppliers.<ResourceHolder<ShardedJedis>>ofInstance(
+          StupidResourceHolder.create(new ShardedJedis(config.getShardsInfo(), Hashing.MURMUR_HASH))
+      );
+    }
+
+    return new RedisCache(clientSupplier, config);
+  }
+
+  public RedisCache(Supplier<ResourceHolder<ShardedJedis>> supplier, RedisCacheConfig config)
+  {
+    this.supplier = supplier;
+    prefix = config.getPrefix();
+    expiration = config.getExpiration();
+    transcoder = new RedisTranscoder();
+  }
+
+  @Override
+  public byte[] get(NamedKey key)
+  {
+    byte[] result = null;
+
+    try (ResourceHolder<ShardedJedis> clientHolder = supplier.get()) {
+      final ShardedJedis jedis = clientHolder.get();
+      final byte[] itemKey = CacheImplUtils.computeKeyHash(prefix, key).getBytes();
+      final byte[] serialized = jedis.get(itemKey);
+
+      if (serialized != null) {
+        result = deserialize(key, serialized);
+        hitCount.incrementAndGet();
+      }
+      else {
+        missCount.incrementAndGet();
+      }
+    }
+    catch (IOException ex) {
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisConnectionException ex) {
+      if (isTimeout(ex)) {
+        timeoutCount.incrementAndGet();
+      }
+
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisException | LZ4Exception ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Error getting cache item");
+      return null;
+    }
+
+    return result;
+  }
+
+  @Override
+  public void put(NamedKey key, byte[] value)
+  {
+    final byte[] serialized;
+    try {
+      serialized = serialize(key, value);
+    }
+    catch (NullPointerException ex) {
+      throw Throwables.propagate(ex);
+    }
+
+    try (ResourceHolder<ShardedJedis> clientHolder = supplier.get()) {
+      final ShardedJedis jedis = clientHolder.get();
+
+      final byte[] itemKey = CacheImplUtils.computeKeyHash(prefix, key).getBytes();
+      Jedis shard = jedis.getShard(itemKey);
```

Comment:
I do this, because occasionally `ShardedJedis` does not implement `psetex()` method.